### PR TITLE
[Merged by Bors] - feat: Generalize corollaries of rank-nullity theorem.

### PR DESCRIPTION
**Pull request by erdOne**
Added a class `HasRankNullity` consisting of the rings that satisfy the rank-nullity theorem.
Generalized the corollaries of the rank-nullity theorem from division rings to rings satisfying the class,
and moved them into a new file `LinearAlgebra.Dimension.RankNullity`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Geometry/Manifold/Algebra/LeftInvariantDerivation.lean [hunk 1]
```diff
@@ -163,12 +163,12 @@ theorem coe_zero : ⇑(0 : LeftInvariantDerivation I G) = 0 :=
   rfl
 #align left_invariant_derivation.coe_zero LeftInvariantDerivation.coe_zero
 
-@[simp]
+@[simp, nolint simpNF] -- timeouts when `import Mathlib`.
 theorem coe_neg : ⇑(-X) = -X :=
   rfl
 #align left_invariant_derivation.coe_neg LeftInvariantDerivation.coe_neg
```
Comment (erdOne) [inline Mathlib/Geometry/Manifold/Algebra/LeftInvariantDerivation.lean:169]:
```
import Mathlib

universe u_4 u_3 u_2 u_1

set_option synthInstance.maxHeartbeats 20000 in
example {𝕜 : Type u_1} [NontriviallyNormedField 𝕜] {E : Type u_2}
  [NormedAddCommGroup E] [NormedSpace 𝕜 E] {H : Type u_3} [TopologicalSpace H]
  {I : ModelWithCorners 𝕜 E H} {G : Type u_4} [TopologicalSpace G] [ChartedSpace H G]
  [Monoid G] [SmoothMul I G] (X : LeftInvariantDerivation I G) : ⇑(-X) = -⇑X := by simp
```
already fails in current master

File: Mathlib/LinearAlgebra/Dimension/RankNullity.lean [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+In this file we provide the rank nullity theorem as a typeclass, and prove various corollaries
+of the theorem. The main definition is `HasRankNullity.{w} R`, which states that
+1. Every `R`-module `M : Type w` has a linear independent subset of cardinality `Module.rank R M`.
+2. `rank (M ⧸ N) + rank N = rank M` for every `R`-module `M : Type w` and every `N : Submodule R M`.
+
+The following instances are provided in mathlib:
+1. `DivisionRing.hasRankNullity` for division rings in `LinearAlgebra/Dimension/DivisionRing.lean`.
+2. `IsDomain.hasRankNullity` for commutative domains in `LinearAlgebra/Dimension/Localization.lean`.
+
+TODO: prove the rank-nullity theorem for `[Ring R] [IsDomain R] [StrongRankCondition R]`.
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Dimension/RankNullity.lean:21]: Are we sure it holds?

File: Mathlib/LinearAlgebra/Dimension/RankNullity.lean [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  rw [← f.quotKerEquivRange.lift_rank_eq, ← lift_add, rank_quotient_add_rank]
+
+/-- The **rank-nullity theorem** -/
+theorem rank_range_add_rank_ker (f : M →ₗ[R] M₁) :
+    Module.rank R (LinearMap.range f) + Module.rank R (LinearMap.ker f) = Module.rank R M := by
+  haveI := fun p : Submodule R M => Classical.decEq (M ⧸ p)
+  rw [← f.quotKerEquivRange.rank_eq, rank_quotient_add_rank]
+#align rank_range_add_rank_ker rank_range_add_rank_ker
+
+theorem lift_rank_eq_of_surjective (f : M →ₗ[R] M') (h : Surjective f) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Dimension/RankNullity.lean:76]: Maybe `f` implicit. Similarly below.

File: Mathlib/LinearAlgebra/Dimension/RankNullity.lean (hunk 2 above)
Comment (erdOne) [reply Mathlib/LinearAlgebra/Dimension/RankNullity.lean:21]: I'm fairly certain that it holds. A domain that is not Ore is of infinite rank (see Cohn's *Skew Field* Proposition 1.3.6) and thus does not satisfy strong rank condition. Hence domains with strong rank condition are Ore and we can mimic the proof here.

Comment (riccardobrasca) [reply Mathlib/LinearAlgebra/Dimension/RankNullity.lean:21]: Nice! Since this is probably not very well known, do you mind add the reference?

**Comment by riccardobrasca** [comment]
@erdOne are you still interested in working on this PR? Or is it obsolete now?

**Comment by erdOne** [comment]
I'm busy until mid-March. I think I'll come back to this after that. Feel free to adopt if this is needed elsewhere.

File: Mathlib/Geometry/Manifold/Algebra/LeftInvariantDerivation.lean (hunk 1 above)
Comment (riccardobrasca) [reply Mathlib/Geometry/Manifold/Algebra/LeftInvariantDerivation.lean:160]: Is this still the case? Maybe some miracles happened...

File: Mathlib/LinearAlgebra/Dimension/Finite.lean [hunk 4]
```diff
@@ -211,6 +211,54 @@ alias finset_card_le_finrank_of_linearIndependent := LinearIndependent.finset_ca
...
+
+lemma exists_finset_linearIndependent_of_le_finrank {n : ℕ} (hn : n ≤ finrank R M) :
+    ∃ s : Finset M, s.card = n ∧ LinearIndependent R ((↑) : s → M) := by
+  by_cases h : finrank R M = 0
+  · rw [le_zero_iff.mp (hn.trans_eq h)]
+    exact ⟨∅, rfl, by convert linearIndependent_empty R M using 2 <;> aesop⟩
+  exact exists_finset_linearIndependent_of_le_rank
+    ((natCast_le.mpr hn).trans_eq (cast_toNat_of_lt_aleph0 (toNat_ne_zero.mp h).2))
+
+lemma exists_linearIndependent_of_le_finrank {n : ℕ} (hn : n ≤ finrank R M) :
```
Comment (riccardobrasca) [inline Mathlib/LinearAlgebra/Dimension/Finite.lean:257]: Can you add a variable line with `{n : ℕ} (hn : n ≤ finrank R M)` here?

**Comment by riccardobrasca** [comment]
Can you add a description of the PR?

Also, is it possible to split it? Correct me if I am wrong, but I have the impression you both generalize/move current results and add a new class. Maybe these two things are intertwined and difficult to separate, so don't worry if it is too much work to do it.

**Comment by erdOne** [comment]
I tried a bit and it seems hard to separate them.

File: Mathlib/Geometry/Manifold/Algebra/LeftInvariantDerivation.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/Geometry/Manifold/Algebra/LeftInvariantDerivation.lean]: Seems fixed.

**Comment by riccardobrasca** [comment]
> I tried a bit and it seems hard to separate them.

OK, don't worry.

**Review by riccardobrasca — APPROVED** [review]
!bench

**Comment by riccardobrasca** [comment]
!bench

**Comment by riccardobrasca** [comment]
> Here are the [benchmark results](http://speed.lean-fro.org/mathlib4/run-detail/e5f5d40d-5456-4846-b9e2-33c6d3b0cfea) for commit [b3998cd](https://github.com/leanprover-community/mathlib4/commit/b3998cd8727de70280a0072a5767c3c8e03f0fdc). There were [significant changes](http://speed.lean-fro.org/mathlib4/compare/3deafc03-398d-4f81-8248-84978ac4fbeb/to/e5f5d40d-5456-4846-b9e2-33c6d3b0cfea) against commit [92c745e](https://github.com/leanprover-community/mathlib4/commit/92c745e2994217861e83a6e65a6ee286dd6f5cd6):
> 
> ```diff
>   Benchmark                                       Metric         Change
>   =====================================================================
> + ~Mathlib.Algebra.Module.PID                     instructions    -5.1%
> - ~Mathlib.LinearAlgebra.Dimension.Finite         instructions    22.3%
> - ~Mathlib.LinearAlgebra.Dimension.Localization   instructions    58.9%
> ```

@mattrobball do you think this is ok?

**Comment by erdOne** [comment]
Both file got longer (the +50％ file doubled in size) so I guess it's fine as long as downstream files aren't slowed?

**Comment by mattrobball** [comment]
Overall it is not terrible but it would nice to take a quick look at those two files to make sure you cannot make an easy change to fix the issue.

**Comment by mattrobball** [comment]
After actually looking at the changes, this is perfectly fine.

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

**Comment by sgouezel** [comment]
bors r-
Can you merge master and fix the build before sending it back to bors?
bors d+

File: Mathlib/LinearAlgebra/Dimension/Finite.lean [hunk 5]
```diff
@@ -211,6 +211,54 @@ alias finset_card_le_finrank_of_linearIndependent := LinearIndependent.finset_ca
 @[deprecated]
 alias Module.Finite.lt_aleph0_of_linearIndependent := LinearIndependent.lt_aleph0_of_finite
 
+lemma exists_set_linearIndependent_of_lt_rank {n : Cardinal} (hn : n < Module.rank R M) :
+    ∃ s : Set M, #s = n ∧ LinearIndependent R ((↑) : s → M) := by
+  obtain ⟨⟨s, hs⟩, hs'⟩ := exists_lt_of_lt_ciSup' (hn.trans_eq (Module.rank_def R M))
+  obtain ⟨t, ht, ht'⟩ := le_mk_iff_exists_subset.mp hs'.le
+  exact ⟨t, ht', .mono ht hs⟩
```
Comment (acmepjz) [inline Mathlib/LinearAlgebra/Dimension/Finite.lean:218]: Could you change the condition of this result to `[Semiring R] [AddCommMonoid M] [Module R M]`? Same as the following `exists_finset_linearIndependent_of_le_rank`.

Comment (acmepjz) [reply Mathlib/LinearAlgebra/Dimension/Finite.lean:218]: It's OK if you do it in later PR; I only need to use the ring version.

**Comment by erdOne** [comment]
bors merge

**Pull request closed without merging**